server/proof_fetcher_test.go

```diff
@@ -0,0 +1,360 @@
+package server
+
+import (
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"reflect"
+	"testing"
+
+	"github.com/golang/mock/gomock"
+	"github.com/golang/protobuf/proto"
+	"github.com/google/trillian"
+	"github.com/google/trillian/crypto"
+	"github.com/google/trillian/merkle"
+	"github.com/google/trillian/storage"
+	"github.com/google/trillian/storage/testonly"
+)
+
+// rehashTest encapsulates one test case for the rehasher in isolation. Input data like the storage
+// hashes and revisions can be arbitrary but the nodes should have distinct values
+type rehashTest struct {
+	desc    string
+	index   int64
+	nodes   []storage.Node
+	fetches []merkle.NodeFetch
+	output  trillian.Proof
+}
+
+// Raw hashes for dummy storage nodes
+var h1 = th.HashLeaf([]byte("Hash 1"))
+var h2 = th.HashLeaf([]byte("Hash 2"))
+var h3 = th.HashLeaf([]byte("Hash 3"))
+var h4 = th.HashLeaf([]byte("Hash 4"))
+var h5 = th.HashLeaf([]byte("Hash 5"))
+
+// And the dummy nodes themselves
+var sn1 = storage.Node{NodeID: storage.NewNodeIDFromHash(h1), Hash: h1, NodeRevision: 11}
+var sn2 = storage.Node{NodeID: storage.NewNodeIDFromHash(h2), Hash: h2, NodeRevision: 22}
+var sn3 = storage.Node{NodeID: storage.NewNodeIDFromHash(h3), Hash: h3, NodeRevision: 33}
+var sn4 = storage.Node{NodeID: storage.NewNodeIDFromHash(h4), Hash: h4, NodeRevision: 44}
+var sn5 = storage.Node{NodeID: storage.NewNodeIDFromHash(h5), Hash: h5, NodeRevision: 55}
+
+// And the output proof nodes expected for them if they are passed through without rehashing
+var n1 = &trillian.Node{NodeHash: h1, NodeId: mustMarshalNodeID(sn1.NodeID), NodeRevision: sn1.NodeRevision}
+var n2 = &trillian.Node{NodeHash: h2, NodeId: mustMarshalNodeID(sn2.NodeID), NodeRevision: sn2.NodeRevision}
+var n3 = &trillian.Node{NodeHash: h3, NodeId: mustMarshalNodeID(sn3.NodeID), NodeRevision: sn3.NodeRevision}
+var n4 = &trillian.Node{NodeHash: h4, NodeId: mustMarshalNodeID(sn4.NodeID), NodeRevision: sn4.NodeRevision}
+var n5 = &trillian.Node{NodeHash: h5, NodeId: mustMarshalNodeID(sn5.NodeID), NodeRevision: sn5.NodeRevision}
+
+// Nodes containing composite hashes. They don't have node ids or revisions as they're recomputed
+var n1n2 = &trillian.Node{NodeHash: th.HashChildren(h2, h1)}
+var n2n3 = &trillian.Node{NodeHash: th.HashChildren(h3, h2)}
+var n2n3n4 = &trillian.Node{NodeHash: th.HashChildren(h4, th.HashChildren(h3, h2))}
+var n4n5 = &trillian.Node{NodeHash: th.HashChildren(h5, h4)}
+
+var rehashTests = []rehashTest{
```

Comment:
Done. Still not convinced that mixing a big block of data into code helps readability.